### PR TITLE
Fix/unique worker names

### LINKED ISSUE #222: `anaconda worker run` cannot run registered worker
If a worker is registered with `--hostname`, there is no way to execute that worker, since the available workers are filtered by the actual hostname of the machine.

Worker names should be unique, so `anaconda worker run <name>` is unambiguous.

hostname should not be filtered, and a warning printed when running on a different hostname/platform.

### PR DESCRIPTION
**Pull request by PeterDSteinberg**
Fixes #222 .

The `anaconda worker list` still filters based on hostname but you are allowed to run a worker without a matching hostname (warning arises then).  Prevents duplicate worker names rather than just preventing duplicate names within one hostname as it was done before.

### CODE REVIEW

File: binstar_build_client/worker/register.py [hunk 1]
```diff
@@ -91,7 +90,7 @@ def __eq__(self, other):
         return self.to_dict() == other.to_dict()
 
     @classmethod
-    def registered_workers(cls, bs):
+    def registered_workers(cls, bs, this_host=False):
```
Comment (dsludwig) [inline binstar_build_client/worker/register.py]: I think this function should always return all `registered_workers`

File: binstar_build_client/worker/register.py [hunk 2]
```diff
@@ -190,7 +195,7 @@ def print_registered_workers(cls, bs):
 
         log.info('Registered workers:')
 
-        for wconfig in cls.registered_workers(bs):
+        for wconfig in cls.registered_workers(bs, this_host=True):
```
Comment (dsludwig) [inline binstar_build_client/worker/register.py]: The body of this function `print_registered_workers` should move to `worker_commands/list.py` - we can add arguments to that command to filter the results, if the user wants to.

File: binstar_build_client/worker/register.py [hunk 3]
```diff
@@ -208,9 +213,9 @@ def register(cls, bs, username, queue, platform, hostname, dist, name=None):
         Register the worker with anaconda server
         '''
         for worker in cls.registered_workers(bs):
-            if worker.name == name:
+            if name in (worker.name, worker.worker_id):
                 raise errors.BinstarError('Cannot have duplicate worker '
-                                          '--name from same host: {}'.format(name))
+                                          '--name or id: {}'.format(name))
```
Comment (dsludwig) [inline binstar_build_client/worker/register.py]: This is good - do we want a server-side check as well?

Comment (dsludwig) [inline binstar_build_client/worker/register.py]: Do we have any issues with visibility still allowing ambiguity?

* John administers Example and Test organizations
* Mary administers Example
* George administers Test
* Mary creates worker-1
* George creates worker-1 (he can't see Example/worker-1)
* John executes `anaconda worker run worker-1` - what happens?

File: binstar_build_client/worker/register.py [hunk 4]
```diff
@@ -174,13 +174,18 @@ def running(self):
...
         for worker in cls.registered_workers(bs):
-            if worker_name == worker.worker_id or worker_name == worker.name:
-                if worker.hostname == cls.HOSTNAME:
-                    return worker
+            if worker_name in (worker.worker_id, worker.name):
+                if warn and worker.hostname != cls.HOSTNAME:
+                    log.warn('Proceeding with worker id registered for '
+                             'different hostname: {}. '
+                             'This host is: {}.'.format(worker.hostname,
+                                                       cls.HOSTNAME))
```
Comment (dsludwig) [inline binstar_build_client/worker/register.py]: I think it is better to duplicate this check and warning in each of `docker_run` and `run` rather than include this check here - it is a confusing api to have this keyword argument `warn`.

File: binstar_build_client/worker/register.py (hunk 3 above)
Comment (PeterDSteinberg) [inline binstar_build_client/worker/register.py]: Great point, @dsludwig .  I had not considered this case yet.  One thing we could do for John in the last step you mention is loop over all the worker names and detect whether there are duplicate names.  If there are duplicate names, e.g. `worker-1` and John tries `anaconda worker run worker-1`, then we could detect there are two possible `worker-1`'s and prompt the user to use one of the corresponding worker id's instead.  (A run time check at start of `anaconda worker run` prints out the worker id's that could be used to disambiguate then it exits non-zero).

It would be nice to do this PR without another change in Anaconda Server.

Comment (dsludwig) [inline binstar_build_client/worker/register.py]: @PeterDSteinberg that sounds like a good solution.

**Comment by PeterDSteinberg** [comment]
The latest commits fix a test, add  a test, and address the edge case mentioned in comments above.  @dsludwig can you take a look at this PR again and also [this docs.anaconda.org PR 257](https://github.com/Anaconda-Server/docs.anaconda.org/pull/257) we should merge with it

File: binstar_build_client/worker/register.py [hunk 5]
```diff
@@ -91,6 +91,25 @@ def __eq__(self, other):
...
 
     @classmethod
+    def validate_worker_name(cls, bs, name):
+        worker_name_to_id = {}
+        for worker in cls.registered_workers(bs):
+            if not worker.name in worker_name_to_id:
+                worker_name_to_id[worker.name] = [worker]
+            else:
+                worker_name_to_id[worker.name].append(worker)
+        worker_name_to_id = {k:v for k,v in worker_name_to_id.items() if len(v) > 1}
```
Comment (dsludwig) [inline binstar_build_client/worker/register.py]: I might move this check `len(workers) > 1` to the if check, rather than processing the entire dictionary. Otherwise the data structure is something like `duplicate_workers_by_name`, which is a bit subtle. Also, it's not `worker_name_to_id`, the value you are storing is a `WorkerConfiguration`, not an `id`.

```
workers = workers_by_name.get(name, [])
if len(workers) > 1:
  # ...
```

**Comment by dsludwig** [comment]
looks good!
